felix/generictables/nft_renderer.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2016-2024 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package generictables
+
+import (
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/knftables"
+
+	"github.com/projectcalico/calico/felix/environment"
+)
+
+const maxNftablesCommentLength = 128
+
+func NewNFTRenderer(hashCommentPrefix string, ipv uint8) NFTRenderer {
+	return &nftRenderer{
+		ipv:               ipv,
+		hashCommentPrefix: hashCommentPrefix,
+	}
+}
+
+type nftRenderer struct {
+	ipv               uint8
+	hashCommentPrefix string
+}
+
+func (r *nftRenderer) Render(chain string, hash string, rule Rule, features *environment.Features) *knftables.Rule {
+	return &knftables.Rule{
+		Chain:   chain,
+		Rule:    r.renderRule(&rule, features),
+		Comment: r.comment(hash, rule),
+	}
+}
+
+func (r *nftRenderer) RuleHashes(c *Chain, features *environment.Features) []string {
+	rf := func(rule *Rule, chain string, features *environment.Features) string {
+		return r.renderRule(rule, features)
+	}
+	return ruleHashes(c, rf, features)
+}
+
+func (r *nftRenderer) renderRule(rule *Rule, features *environment.Features) string {
+	fragments := []string{}
+
+	if rule.Match != nil {
+		matchFragment := rule.Match.(NFTMatchCriteria).IPVersion(r.ipv).Render()
+		if matchFragment != "" {
+			fragments = append(fragments, matchFragment)
+		}
+	}
+
+	if rule.Action != nil {
+		// Include a counter action on all rules.
```

Comment:
Yeah, we probably do. For now, having a counter on each rule has helped me to diagnose subtle bugs since it's a bit easier to track exactly where packets are going.